packages/contracts-bedrock/contracts/L2/CrossDomainOwnable3.sol

```diff
@@ -0,0 +1,67 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { Predeploys } from "../libraries/Predeploys.sol";
+import { L2CrossDomainMessenger } from "./L2CrossDomainMessenger.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+
+/**
+ * @title CrossDomainOwnable2
+ * @notice This contract extends the OpenZeppelin `Ownable` contract for L2 contracts to be owned
+ *         by contracts on L1. Note that this contract is meant to be used with systems that use
+ *         the CrossDomainMessenger system. It will not work if the OptimismPortal is used
+ *         directly.
+ */
+abstract contract CrossDomainOwnable3 is Ownable {
+    /**
+     * @notice If true, the contract uses the cross domain _checkOwner function override. If false
+     *         it uses the standard Ownable _checkOwner function.
+     */
+    bool internal isLocal = true;
+
+    event OwnershipTransferred(
+        address indexed previousOwner,
+        address indexed newOwner,
+        bool isLocal
+    );
+
+    /**
+     * @notice Overrides the implementation of the `onlyOwner` modifier to check that the unaliased
+     *         `xDomainMessageSender` is the owner of the contract. This value is set to the caller
+     *         of the L1CrossDomainMessenger.
+     */
+    function _checkOwner() internal view override {
+        if (isLocal) {
+            require(owner() == _msgSender(), "CrossDomainOwnable3: caller is not the owner");
+        } else {
+            L2CrossDomainMessenger messenger = L2CrossDomainMessenger(
+                Predeploys.L2_CROSS_DOMAIN_MESSENGER
+            );
+
+            require(
+                msg.sender == address(messenger),
+                "CrossDomainOwnable3: caller is not the messenger"
+            );
+
+            require(
+                owner() == messenger.xDomainMessageSender(),
+                "CrossDomainOwnable3: caller is not the owner"
+            );
+        }
+    }
+
+    /**
+     * @notice Overrides the implementation of the `transferOwnership` function to allow
+     * for local ownership.
+     * @param _owner The new owner of the contract.
+     * @param _isLocal If false, the contract uses the cross domain _checkOwner function override.
+     * If false it uses the standard Ownable _checkOwner function.
+     */
+    function transferOwnership(address _owner, bool _isLocal) external {
```

Comment:
We want a test that this reverts when not the owner calls it

Reply:
Because of the revert message not being prefixed with `CrossDomainOwnable3`, we may as well pull the impl of `transferOwnership` inside of here and use the proper revert message

Reply:
will add it!

Reply:
I had to rename `_owner` back to `newOwner` because `_owner` is the internal variable that holds the owner in `Ownable.sol`. I think that's cleaner than having other contract naming conventions takeover. Especially if we move `super.transferOwnership` functionality into the contract.

Reply:
Ok, after some trial and error, I did leave `super.transferOwnership` because it does use `onlyOwner` where we redefine `_checkOwner` to have `CrossDomainOwnable3:` as a prefix for errors. I only added the require statement that the new owner is not the zero address.